### PR TITLE
gtest: Add support for Conan V2

### PR DESCRIPTION
**Pull request by jwillikers**
Specify library name and version:  **gtest/***

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by jwillikers** [comment]
Fails to build shared due to issue conan-io/conan#11840.

**Comment by uilianries** [comment]
@jwillikers You can use same workaround: https://github.com/conan-io/conan-center-index/pull/12110/files#diff-0d190e2175005f8bc838045b75be04ba9b92627a199259d8b056d7a8f236a858R59

**Comment by jwillikers** [comment]
@uilianries Thanks!

**Comment by uilianries** [comment]
I just re-triggered again. I could be a runtime error, that's why we don't see a message error.

**Comment by uilianries** [comment]
It failed again, but no clue about.

**Comment by jwillikers** [comment]
> It failed again, but no clue about.

@uilianries I fixed that issue, but now I've hit a weird one both here and in updates to the `gtest` Conan package in PR #12411.
There's a link error in the test package step when building with Clang and `libstdc++`.
I'm able to reproduce the link error locally with Conan 1.51.3, too.

File: recipes/gtest/all/conanfile.py [hunk 1]
```diff
@@ -112,57 +104,65 @@ def loose_lt_semver(v1, v2):
...
-        cmake = CMake(self)
+    def generate(self):
+        tc = CMakeToolchain(self)
+
+        # Honor BUILD_SHARED_LIBS from conan_toolchain (see https://github.com/conan-io/conan/issues/11840)
+        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0077"] = "NEW"
+        
         if self.settings.build_type == "Debug":
-            cmake.definitions["CUSTOM_DEBUG_POSTFIX"] = self.options.debug_postfix
+            tc.cache_variables["CUSTOM_DEBUG_POSTFIX"] = str(self.options.debug_postfix)
```
Comment (uilianries) [inline recipes/gtest/all/conanfile.py:132]:
```suggestion
            tc.variables["CUSTOM_DEBUG_POSTFIX"] = str(self.options.debug_postfix)
```

File: recipes/gtest/all/conanfile.py [hunk 2]
```diff
@@ -112,57 +104,65 @@ def loose_lt_semver(v1, v2):
...
         if is_msvc(self) or self._is_clang_cl:
-            cmake.definitions["gtest_force_shared_crt"] = "MD" in msvc_runtime_flag(self)
-        cmake.definitions["BUILD_GMOCK"] = self.options.build_gmock
+            tc.cache_variables["gtest_force_shared_crt"] = "MD" in msvc_runtime_flag(self)
```
Comment (uilianries) [inline recipes/gtest/all/conanfile.py:134]:
```suggestion
            tc.variables["gtest_force_shared_crt"] = "MD" in msvc_runtime_flag(self)
```

File: recipes/gtest/all/conanfile.py [hunk 3]
```diff
@@ -112,57 +104,65 @@ def loose_lt_semver(v1, v2):
...
+
+        try:
+           check_min_vs(self, "191")
+        except ConanInvalidConfiguration:
+            tc.preprocessor_definitions["GTEST_LANG_CXX11"] = 1
+            tc.preprocessor_definitions["GTEST_HAS_TR1_TUPLE"] = 0
+            
+        tc.cache_variables["BUILD_GMOCK"] = bool(self.options.build_gmock)
```
Comment (uilianries) [inline recipes/gtest/all/conanfile.py:142]:
```suggestion
        tc.variables["BUILD_GMOCK"] = bool(self.options.build_gmock)
```

File: recipes/gtest/all/conanfile.py [hunk 4]
```diff
@@ -112,57 +104,65 @@ def loose_lt_semver(v1, v2):
...
         if self.settings.os == "Windows" and self.settings.compiler == "gcc":
-            cmake.definitions["gtest_disable_pthreads"] = True
-        cmake.definitions["gtest_hide_internal_symbols"] = self.options.hide_symbols
-        cmake.configure(build_folder=self._build_subfolder)
-        return cmake
+            tc.cache_variables["gtest_disable_pthreads"] = True
+        tc.cache_variables["gtest_hide_internal_symbols"] = bool(self.options.hide_symbols)
```
Comment (uilianries) [inline recipes/gtest/all/conanfile.py:145]:
```suggestion
            tc.variables["gtest_disable_pthreads"] = True
        tc.variables["gtest_hide_internal_symbols"] = bool(self.options.hide_symbols)
```

**Review by uilianries — COMMENTED** [review]
cache_variables is not needed in most cases.

**Comment by uilianries** [comment]
> > It failed again, but no clue about.
> 
> @uilianries I fixed that issue, but now I've hit a weird one both here and in updates to the `gtest` Conan package in PR #12411. There's a link error in the test package step when building with Clang and `libstdc++`. I'm able to reproduce the link error locally with Conan 1.51.3, too.

Which error? Log?

File: recipes/gtest/all/conanfile.py (hunk 4 above)
Comment (jwillikers) [reply recipes/gtest/all/conanfile.py]: Done.

File: recipes/gtest/all/conanfile.py (hunk 2 above)
Comment (jwillikers) [reply recipes/gtest/all/conanfile.py]: Done.

File: recipes/gtest/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/gtest/all/conanfile.py]: Done.

**Review by uilianries — DISMISSED** [review]

**Review by uilianries — APPROVED** [review]

**Review by toge — APPROVED** [review]

**Pull request merged**